Update Appendix A for finance-quote 1.63

pkryger opened the pull request:
A few weeks ago @bpschuck [asked for help](https://github.com/finance-quote/finance-quote/discussions/423) with updating `Gnucash` for recent version of `finance-quote`. This is a complitentary PR to https://github.com/Gnucash/gnucash/pull/2033.

I decided to move modules that are not a part of the list into a new section for deprecated sources. IIUC this is still usefull information for a case when an old `.gnucash` file (or a database) has been configured with one of the deprecated sources.

pkryger committed:
Update Appendix A for finance-quote 1.63

pkryger committed:
Add back cse - Colobmo Stock Exchange

jralls commented on C/manual/tips-appendix.docbook:
```diff
@@ -6457,4 +6316,662 @@ bogus_symbol2,price2,date2
     </para>
   </sect1>
+  <sect1 id="fq-deprecated-sources">
+    <title>Deprecated Finance::Quote Sources</title>
+
```
Deprecated identifies something that still works but will be eventually removed with eventually meaning some time between the next release and the heat death of the universe.

Obsolete would be better.

fellen replied:
Oh yes, please.

fellen replied:
And IMHO the structure should be:
```
<sect1…><title>Finance::Quote Sources</title>…
  <sect2 …> <title>Quote Sources - Individual Sources</title>
    <sect3 …> <title>Active Individual Sources</title>
    <sect3 …><title>Obsolete Individual Sources</title>
  <sect2 …><title>Quote Sources - Multiple Sources</title>
    <sect3 …><title>Active Multiple Sources</title>
    <sect3 …><title>Obsolete Multiple Sources</title>
<sect1…><title>Details of Some Sources</title>…
 <sect2…><title>Yahoo Specifics</title>…
 <sect2…><title>TIAA-CREF Specifics</title>…
```
Less  scrolling for the users searching their old source.

pkryger replied:
> [...] Obsolete would be better.

Updated. Added a bit of abstract too.

> And IMHO the structure should be: [...]

Updated.  It's nicer to my brain indeed. Thanks!

jralls commented:
I think this needs to be in the wiki instead of the manual: It changes a lot and we're terrible at keeping it up to date.

fellen commented:
> I think this needs to be in the wiki instead of the manual: It changes a lot and we're terrible at keeping it up to date.

I had created this appendix before we started the wiki.
It is easy done, while updating engine/gnc-commodity.cpp like in  https://github.com/Gnucash/gnucash/pull/2033.

OTOH one could think about adding more infos like the URL, API key requirement, … already in the documentation of the perl sources and populate corresponding fields in QuoteSourceList.

jralls commented:
> It is easy done, while updating engine/gnc-commodity.cpp like in https://github.com/Gnucash/gnucash/pull/2033.

We haven't done a great job of keeping gnc-commodity.cpp up to date either.

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -246,122 +232,119 @@
             <row>
               <entry>
-                <para>BAMOSZ funds, HU
+                <para>Bloomberg
```
Region missing, usually the TLD of the URL, mapping com, mil, … to us.
Perhaps we should add a term like `multi ` or `supraregional` to the global players?

pkryger replied:
Sorry, I forgot to replay earlier.

 I intentionally left the region out here and in a few other places (`marketwatch`, `stockdata`, `twelvedata`). This was to follow what has been done in `financeapi` and, to some degree, `goldmoney`. I guess adding some extra term here should be followed in other places, right?

As for using term `multi`: I'd refrain from doing that, as there are already "Multiple Quote Sources" and this may cause confusion. `global` perhaps?

jralls replied:
The TLD isn't useful, there are too many supposedly country TLDs that have been commercialized. `supranational` seems a bit legalistic and `global` implies a bit wider scope than is the case. How about `multnational`, a term widely used in the US for companies doing business in more than one country?

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -786,68 +727,83 @@
 
               <entry>
-                <para>tdefunds
+                <para>torouro_direto
```
Typo?[Edit:] tesouro_direto

pkryger replied:
great catch!

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -402,193 +385,151 @@
-              <entry>
-                <para><ulink url="https://www.goldmoney.com"/>
+                    <listitem>
+                      <para>Export the key as <envar>IEXCLOUD_API_KEY</envar> environment variable before starting &appname;.
```
@jralls I wonder if we should add storages for this new envars in preferences.

pkryger replied:
Let me know when that happens and I will update this document.

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -876,63 +832,78 @@
               <entry>
-                <para><ulink url="https://www.trustnet.com"/>
+                <para>You need to
+                  <orderedlist spacing="compact">
```
[`<procedure>`](https://tdg.docbook.org/tdg/4.5/procedure) is better than `<orderedlist>`here.

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -6457,4 +6316,662 @@ bogus_symbol2,price2,date2
+                <para><ulink url="https://www.cominvest-am.de"/>
+                </para>
+
+                <para>Obsolete, update: <ulink url="https://eggert.org/software/Comdirect.pm"/>
```
Perhaps the FQ team could, if still usable, merge his version and we can remove this note?

pkryger replied:
Updated the note. `comdirect` is alrady a part of `Finance::Quote`

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -6457,4 +6316,662 @@ bogus_symbol2,price2,date2
+
+              <entry>
+                <para><ulink url="https://www.six-swiss-exchange.com"/>
+                </para>
```
This and the next were joined to SIX.
Perhaps do it like in Amsterdam Euronext eXchange, NL?

pkryger replied:
Done

fellen reviewed (commented):
A very good start! Please check my notes.

[Edit:] Should we add version infos like `introduced in FQ version x.y`? 
Perhaps we should add classifications like `They offer shares and/or funds and/or ForEx, Indices, Options, …`

pkryger committed:
Change deprecated to obsolete for quote sources

pkryger committed:
Make Active and Obsolete Finance::Quote Sources into a single section

pkryger committed:
Fix typo in tesouro_direto

pkryger committed:
Capitalise Currency Source in title

pkryger committed:
Use procedure instead of orderedlist for Finance::Quotes API Keys

pkryger committed:
Update comment about comdirect

pkryger committed:
Update comment for SIX

pkryger commented:
@jralls @fellen thank you for taking a time and reviewing this.

I applied most of the suggestions from comments (and resolved them). I did that as individual commits to help with spotting what has actually been changed since the previous read. Note that I also changed a structure, which rendered another large change 🤷‍♂️.

Please let me know if this has reached a state good enough start Wiki port. Also, could you please let me know what's need to be done to perform such a move.

pkryger committed:
Remove ellipses from multiple quote sources

jralls commented:
> Wiki port...

Do you have a wiki account already? If so you can just add the list [here](https://wiki.gnucash.org/wiki/Online_Quotes#About_Finance::Quote). If not you'll have to request an account.

fellen commented:
> Wiki port

To avoid redundancy you should IMHO only update the content of  [`<SyntaxHighlight lang="console">`](https://wiki.gnucash.org/wiki/index.php?title=Online_Quotes&action=edit&section=6)

jralls commented:
> To avoid redundancy you should IMHO only update the content of [<SyntaxHighlight lang="console">](https://wiki.gnucash.org/wiki/index.php?title=Online_Quotes&action=edit&section=6)

Since that's the actual output from `gnucash-cli -Q info`, no.

To avoid redundancy we're going to remove this appendix because it's too hard to keep it up to date.

bpschuck commented:
On 10/19/24 04:20, Przemysław Kryger wrote:

> As for using term |multi|: I'd refrain from doing that, as there are
>  already "Multiple Quote Sources" and this may cause confusion.
> |global| perhaps?

PK,

I believe you are losing sight of what the F::Q code refers to them as, 
"failover" methods. In some ways, these are "virtual" methods. Sort of 
like round-robin DNS load balancing.

Hypothetically, I have FOOB defined to use source "nasdaq". F::Q will 
try to use the various modules where "nasdaq" is advertised. Either one 
of the methods that "nasdaq" is an alias for is successful, or F::Q 
tries all of them (currently 8) an finally returns with a success=0.

A module where the method can also be called using "nyse", "nasdaq", 
"usa", or "amex", isn't "global".

I would change the table to something like the following, after of 
course a quick description of how they are used.

Failover Methods

Failover Method - Possible Source Method
nyse            - alphavantage, financeapi, fool, googleweb, ...
nasdaq          - alphavantage, financeapi, fool, googleweb, ...
india           - bseindia, indiamutual, nseindia
.
.
.

Bruce S.

bpschuck commented:
On 10/19/24 10:38, John Ralls wrote:

> To avoid redundancy we're going to remove this appendix because it's too 
> hard to keep it up to date.

The bane of my existence at the day job. Supporting some legacy 
applications that this was the obvious choice of long gone developers. 
To paraphrase, "we're not going to write documentation because it would 
be too hard to keep up to date"... Hence we've had some outages where we 
had to spend hours reverse engineering before we could identify the 
issue at hand.

Seriously though, once it is pretty much caught up it should be 
relatively easy to keep up to date. As long as there is an effort to. 
But like you said earlier John, perhaps moving it to the GnuCash Wiki is 
a much better alternative.

Then again, it is something that should be more of F::Q doc, and not 
GnuCash. Not exactly something that is fit for POD, and the GitHub 
supplied Wiki isn't exactly one that is full featured. While I did 
recently register finance-quote.org, I'm not going to pay for a hosting 
service out of my own pocket. Hell, I'm not sure how much longer I'll be 
playing F::Q gatekeeper...

Bruce S.

jralls commented:
@bpschuck 
> I believe you are losing sight of what the F::Q code refers to them as, "failover" methods. In some ways, these are "virtual" methods. Sort of like round-robin DNS load balancing.

It's not about that, it's about sources like AlphaVantage and Yahoo_JSON that provide quotes from more than one country. What F::Q calls failover sources are listed in a gnucash array named [`multiple_quote_sources`](https://github.com/Gnucash/gnucash/blob/3d0e65bb7f0a2b40f693d447e4ca62a043513dc1/libgnucash/engine/gnc-commodity.cpp#L223), hence  @pkryger's aversion to "multi" for the likes of AlphaVantage.

> Then again, it is something that should be more of F::Q doc, and not GnuCash.

There's no reason it can't be in both. It could be in POD but few GnuCash users are going to read POD. The Github-provided wiki might be limited compared to MediaWiki, but it's certainly good enough for this purpose. On the other hand you're welcome to use GnuCash's wiki (you've already got an account) and place links to it from Finance::Quote.

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -45,6416 +63,6935 @@
+                </entry>
+
+                <entry>
+                  <para><ulink url="https://www.six-swiss-exchange.com"/>
```
`Part of `
Perhaps replace the URL by `SIX Swiss Exchange, CH` — below and in similar expresions with their GnuCash name also.

pkryger replied:
done

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -45,6416 +63,6935 @@
+                </entry>
+
+                <entry>
+                  <para><ulink url="https://www.six-swiss-exchange.com"/>
```
`Part of `

pkryger replied:
done

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -33,7 +34,24 @@
+
+    <para>If a particular source is the only place where the quote can be obtained but it's been failing or is obsolete,
+    please reach out to <application>Finance::Quote</application> maintainers at
+    <ulink url="https://github.com/finance-quote/finance-quote"/>.
```
Define the url  in docbook/gnc-docbookx.dtd as
`<!ENTITY url-fq "https://github.com/finance-quote/finance-quote">`
and use it here as
`<ulink url="&url-fq;"/>`
That is easier to maintain over all languages.

pkryger replied:
done

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -33,7 +34,24 @@
+    </para>
+
+    <para>Some sources may be considered obsolete that means that support for them has been removed (or is being
+    removed), or they are not functioning correctly. Please see <link linkend='fq-obsolete-sources-single'>Obsolete
```
@bpschuck Should we add a few words about your policy on removing and perhaps reenabling?

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -33,7 +34,24 @@
+
+    <para>Some sources may be considered obsolete that means that support for them has been removed (or is being
+    removed), or they are not functioning correctly. Please see <link linkend='fq-obsolete-sources-single'>Obsolete
+    Single Quote Sources</link> and <link linkend='fq-obsolete-sources-multiple'>Obsolete Multiple Quote Sources</link>
```
For internal links we prefer `<xref linkend='fq-obsolete-sources-multiple'/>` (with autogenerated label).

pkryger replied:
done

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -45,6416 +63,6935 @@
+            </tbody>
+          </tgroup>
+        </table>
+        <para>Sources: src/engine/gnc-commodity.c:gnc_quote_source (commit c0fd3b3, which was adjusted for
```
1. src/ was in betweeen split in GTK based gnucash/ and GUI  independant libgnucash/
2. the  compiler changed from C to C++

`libgnucash/engine/gnc-commodity.cpp`

The commit hash can still change. We should update it after committing the program PR and perhaps use a link then..

pkryger replied:
done

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -45,6416 +63,6935 @@
+
+              <row>
+                <entry>
+                  Europe (ASEGR, Bourso, BVB, Consorsbank, Sinvestor, Stooq, Tradegate, XETRA)
```
Questions: Do they all support the ISIN as symbol or are they using different national IDs like german WKN?
If they use ticker symbols are they defined somewhere globally or has each market its own list?
[A few Examples](https://wiki.gnucash.org/wiki/Online_Quotes#Setting_up_online_quotes_in_GnuCash)

fellen reviewed (commented):
Today I found a few minor issues — mostly from  [our ruleset](https://wiki.gnucash.org/wiki/Documentation_Improvement#Optimize_the_Compression).

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -45,6416 +63,6935 @@
+
+                <entry>
+                  <note>
+                    <para>You need to</para>
+                    <procedure>
```
At least Ubuntu seems to have problems. How about

1. Upgrade `<note>` to `<important>`,
2. Replace the introducting `<para>` by a `<title>`
`<procedure><title>API Key Required</title>`

pkryger replied:
done

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -45,6416 +63,6935 @@
+                </entry>
+
+                <entry>
+                  <para><ulink url="https://www.cominvest-am.de"/>
+                  </para>
+
```
Remove it, because
Die Verbindung mit dem Server www.cominvest-am.de schlug fehl.

pkryger replied:
done

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -45,6416 +63,6935 @@
+                  <para><ulink url="https://www.cominvest-am.de"/>
+                  </para>
+
+                  <para>Obsolete, use comdirect"/>
```
`The Commerzbank AG changed their structure, use their brand <emphasis>comdirect</emphasis> now.`

pkryger replied:
done

pkryger committed:
Add finance-quote url

pkryger committed:
Update comments about six

pkryger committed:
Update sources reference

pkryger committed:
Use xref for internal links

pkryger reviewed (commented):
I think the content of this PR is in a reasonable condition to start porting it to GnuCash Wiki. I will close this PR once I am done with the port and I will provide a link from Wiki to here for future archeological excursions.

fellen commented on docbook/gnc-docbookx.dtd:
```diff
@@ -195,6 +195,7 @@ https://www.w3.org/2003/entities/2007/w3centities-f.ent
 <!-- Fixme: In April 2012, the XML project was retired
   For more information, please explore the Attic [http://attic.apache.org/]. -->
 <!ENTITY url-xalan "https://xml.apache.org/"> <!-- XSLT and XQuery processor @ apache -->
+<!ENTITY url-fq "https://github.com/finance-quote/finance-quote">
```
Alphabetical order inside of sections means after url-fints as line 169.

pkryger replied:
done

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -33,7 +34,23 @@
+    </para>
+
+    <para>Some modules are considered Active, that means it should be possible to fetch quotes using them. Please
+    see <link linkend='fq-active-sources-single'>Active Individual Quote Sources</link>
+    and <link linkend='fq-active-sources-multiple'>Active Multiple Quote Sources</link> for detailed lists.
```
like below
```
    see <xref linkend='fq-active-sources-single'/>
    and <xref linkend='fq-active-sources-multiple'/> for detailed lists.
```
is shorter and more fail-safe.

pkryger replied:
done

fellen commented on C/manual/tips-appendix.docbook:
```diff
@@ -45,6416 +62,6935 @@
 
     <sect2 id="fq-currencies">
-      <title>Finance::Quote Sources - Currency source</title>
+      <title>Finance::Quote Sources - Currency Source</title>
```
In previous para add a xref, please:
`      ISO currencies. The other two can be selected in the <xref linkend="tool-security-edit"/>.`
This should already have been done years before.

pkryger replied:
done

fellen reviewed (commented):
Just some more polishing:

pkryger committed:
Add xreff to security editor

pkryger committed:
Remove region from AlphaVantage

pkryger committed:
Use English names for Italian and Bucharest stock exchanges

pkryger commented:
So I exported contents of this chapter into Wiki, please see https://wiki.gnucash.org/wiki/Online_Quotes#Finance::Quote_Sources and https://wiki.gnucash.org/wiki/Online_Quotes#Details_of_Some_Finance::Quote_Sources.

I am closing this PR and will open a new one, to remove this file.

#### The gist of how the export was made

I used automated export with:
``` shell
pandoc -f docbook -t mediawiki ../C/manual/tips-appendix.docbook -o tips-appendix.wiki
```
And later edited it manually to (note: regexps as in Emacs):
* increase nesting level `s/^\(=+\)\( [a-z0-9 :-]+\)\(=+\)"/\1=\2\3=/`,
* converted most of links to inline `s/\[\(https?://.+\)\]/\1/` (note: not all!),
* updated API Keys procedures (manually),
* compressed tables
  * removed empty lines `s/^\([|!].*\)\n\n/\1\n/`
  * converted to single row in markup, where applicable: 
    * first 4 columns: `s/`^\([!|]\)\( .*\)\n\([!|]\)\( .*\)\n\([!|]\)\( .*\)\n\([!|]\)\( .*\)/\1\2 \3\3\4 \5\5\6 \7\7\8/`
    * then 3 columns: `s/^\([!|]\)\( .*\)\n\([!|]\)\( .*\)\n\([!|]\)\( .*\)/\1\2 \3\3\4 \5\5\6/`

pkryger commented:
Closing - this section has been exported to https://wiki.gnucash.org/wiki/Online_Quotes#Finance::Quote_Sources

fellen commented:
Several reasons

1. Thanks to the Great Firewall users in CN can not access wikis.
2. primary the appendix is documenting tables in GnuCash's engine.
3. Until now part of the documentation policy was: the Docu team moves parts from the wiki to docs and replace them by a link. Common users can still add new developments to the wiki. Do we really want to revert that?

jralls commented:
@fellen, knock it off. There is no "Doc Team" and you darn well know it. 

The fact that there are tables for this in GnuCash is an implementation detail that has no business being in the Manual.

The PRC's [blocking of Wikipedia](https://en.wikipedia.org/wiki/Internet_censorship_in_China) is well documented, but that article  says nothing about blocking all wikis. @xuxinhang are you able to access https://wiki.gnucash.org/wiki/Online_Quotes?

Besides, it's pretty much impossible to keep this up to date considering that F::Q is on a different release cadence from GnuCash and the various websites that F::Q depends upon do breaking changes pretty much at random.

fellen commented:
The first column <q>GnuCash Name</q> in https://wiki.gnucash.org/wiki/Online_Quotes#Finance::Quote_Sources only exist after an entry was added in gnc-commodity.cpp.
An wiki update without a change in gnc-commodity.cpp is pointless.

Perhaps we can get a recent useful table by [Doxygen](https://wiki.gnucash.org/wiki/Doxygen)? [File](https://code.gnucash.org/docs/STABLE/gnc-commodity_8cpp_source.html), existing page [ QuoteSourceType](https://code.gnucash.org/docs/STABLE/group__Commodity.html#gabc4109dd1e05729046a0f98c98ebe3b9)

jralls commented:
> Perhaps we can get a recent useful table by [Doxygen](https://wiki.gnucash.org/wiki/Doxygen)? [File](https://code.gnucash.org/docs/STABLE/gnc-commodity_8cpp_source.html), existing page QuoteSourceType

For who? The Doxygen docs are for developers, who care about the [structure and the functions](https://code.gnucash.org/docs/STABLE/structgnc__quote__source__s.html). It's rare that the values would be important for anything other than testing.

xuxinhang commented:
> The PRC's [blocking of Wikipedia](https://en.wikipedia.org/wiki/Internet_censorship_in_China) is well documented, but that article says nothing about blocking all wikis. @xuxinhang are you able to access https://wiki.gnucash.org/wiki/Online_Quotes?


The wiki.gnucash.org is accessible via my ISP.

China's Firewall works on the black-list containing DNS names of some famous companies, so it doesn't care which software the website is hosted on. In fact, most of overseas websites are accessible to Chinese visitors, excluding popular social media and news media websites.
